writer/template/ring.go

```diff
@@ -0,0 +1,257 @@
+package template
```

Comment:
I don't wanna block on this, but I think focusing on this as a `Writer` instead of as a generic ringbuffer that counts wraparaounds, etc makes it harder to think about as a reviewer or someone debugging. Your tests focus on writer as a unit instead of "I added `max` + 1 to the channel, did the ring wrap?" Again, might not matter in the long term but made this take longer for me.